### ISSUE TITLE
tape.gradients returned in wrong order when using jax2tf+tf.Module

### ISSUE THREAD

**Issue by tomhennigan** [google/jax#10586]
Originally reported in deepmind/dm-haiku#389.

A minimal reproducer can be found in this Colab: https://colab.research.google.com/gist/tomhennigan/144e75ca79627cfd0edd98113a90e1c0/jax2tf.ipynb.

I've found that you can "fix" the issue by removing `tf.Module` as the base class of `JaxModule`.

My hypothesis for why this fixes things: `tf.Module` inherits from `Checkpointable`. This means that (despite what the code makes it look like) `self._params` will not actually be a `dict` instance, it will be some `DictWrapper` thing that Checkpointable uses to track mutations. I suspect the issue is that somewhere in JAX or TF we are relying on the iteration order of dicts and it differs between `dict` and `DictWrapper` (not 100% sure on the name).

This wasn't previously an issue (the notebook ran when it was checked in in Nov 2020) so I am not sure if something has changed in JAX or TF.

**Comment by gnecula** [comment]
The problem comes from the fact that `jax.tree_util.flatten` behaves somewhat differently than `tf.nest.flatten` on a `DictWrapper`. I will think if there is a clean solution or simply we should give an error in that case. 

The same would happen if the argument contains a type for which either JAX or TensorFlow register custom pytree handlers. In that case probably the best thing to do is to give an error.
